Add support for CBOR stringref extension.

here-abarany opened the pull request:
See http://cbor.schmorp.de/stringref for stringref definition. Repeated binary and text strings are assigned an index on write and a tagged integer is written when repeated. Parser respects the stringref IDs when reading data back in.

Nested tags are now supported. The previous getCurrentTag() function is still supported for backwards compatibility, and returns the first tag. Otherwise a new TagList type is used to efficiently handle multiple tags.

cowtowncoder commented:
Ok, first of all: thank you for contributing this feature! It sounds a useful thing, CBOR implementing support something Smile has had for a while.

One thing I was wondering about: since this is a sizable thing, I was wondering if it was possible to split this in two PRs: one dealing with parsing, and second one (building on first) for generation? I realize it may be tricky wrt tests, but it would be easier to code review that way if at all possible.

Second thing: regardless of whether it's 1 or 2 PRs, one (and only) process thing we need is CLA.
It can be found from here:
https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

here-abarany commented:
> One thing I was wondering about: since this is a sizable thing, I was wondering if it was possible to split this in two PRs: one dealing with parsing, and second one (building on first) for generation?

From what I understand this would require separate branches, which would be very unwieldy to manage and would require the first PR to be submitted before the second could actually be looked at. (since the second one would contain the code from the first) What I can do is split it into two commits, which would allow you to review them separately while being part of the same branch/PR.

> Second thing: regardless of whether it's 1 or 2 PRs, one (and only) process thing we need is CLA.

Since I'm contributing this on company time I've forwarded this to the OSS department to make sure I get the OK to sign it. Hopefully it won't take too long to get approval.

cowtowncoder commented:
> > One thing I was wondering about: since this is a sizable thing, I was wondering if it was possible to split this in two PRs: one dealing with parsing, and second one (building on first) for generation?
> 
> From what I understand this would require separate branches, which would be very unwieldy to manage and would require the first PR to be submitted before the second could actually be looked at. (since the second one would contain the code from the first) What I can do is split it into two commits, which would allow you to review them separately while being part of the same branch/PR.
> 
> > Second thing: regardless of whether it's 1 or 2 PRs, one (and only) process thing we need is CLA.
> 
> Since I'm contributing this on company time I've forwarded this to the OSS department to make sure I get the OK to sign it. Hopefully it won't take too long to get approval.

Yes, would require 2 branches and I guess that won't play nicely with PRs. Too bad but I guess it makes sense to keep things bundled.

And yes, CLA should be fine; there's CCLA available if your company prefers that.

here-abarany commented:
Just to give a heads up, we have a tentative agreement internally to sign the CCLA, but are trying to find the right people to sign it. I'm hoping we'll get it submitted by mid January. In the meantime I'll be out the next two weeks myself for the holidays.

here-abarany commented:
@cowtowncoder I finally managed to get a signature and forwarded the CCLA last Friday. I just wanted to make sure that it was received and we can continue with the PR.

cowtowncoder commented:
@here-abarany I can't seem to find CCLA. Was it sent to `info` at fasterxml dot com?

here-abarany commented:
@cowtowncoder it was sent to `clas` rather than `info`, which was the address in the document. I can forward it to `info` as well.

cowtowncoder commented:
@here-abarany Ok odd. I think `clas` should get forward to `info`... but yes, please try re-sending it as I don't see it.

here-abarany commented:
@cowtowncoder I just forwarded it now.

cowtowncoder commented on cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:
```diff
@@ -924,8 +958,25 @@ public void writeBinary(Base64Variant b64variant, byte[] data, int offset,
         }
         _verifyValueWrite("write Binary value");
+        ByteBuffer bytesRef = null;
+        if (_stringRefs != null) {
```
Does StringRefs functionality also apply to binary values? That seems.... odd. Not all binary values represent Strings.

Would it make sense to have separate features for enabling these separately? (I assume it is fine to write combination of references and non-referenced String / binary values).

here-abarany replied:
The extension requires computing the index for every binary and text string within a namespace as long as it fits within the minimum size based on the current index value. (i.e. the more bytes required to store the index, the longer the threshold to compute the index for sharing) It doesn't allow you to selectively apply indices with the exception to indefinite length (chunked) strings, which will never have an index computed.

In other words, the only way to exclude binary strings from the index computation is to always write them with the indefinite length encoding.

cowtowncoder replied:
Ok. So it is required by CBOR spec. Interesting choices; esp. not having upper limit. Could be problematic for many binary data cases but I guess that's not my problem :)
(I spec'd/implemented String back refs for Smile format before CBOR was a thing; never even considered shared refs for binary data so this seems.... interesting).

here-abarany replied:
I would imagine that if a client has a significant amount of binary data to the point where this becomes an issue they could simply not opt into stringrefs. If clients are concerned about *both* a significant number of shared strings bloating their data *and* large embedded binaries (or large strings for that matter), an option would be to introduce a configuration to have an upper limit on string size before it forces a chunked encoding. (which would bypass the stringref sharing) I would imagine this is very likely very rare, however, and probably not worth the complexity of introducing an option until there's a real-world example that both proves the need for it and provides an example to more effectively create such a feature.

cowtowncoder commented on cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java:
```diff
@@ -245,6 +322,35 @@ private Feature(boolean defaultState) {
+    /**
+     * Type to keep track of a list of string references. A depth is stored to know when to pop the
+     * references off the stack for nested namespaces.
+     */
```
Add `@since 2.15` for new classes, public methods in public API.

here-abarany replied:
I did a quick check and added it to this and the new tag constants in CBORConstants. Let me know if I missed any others.

cowtowncoder replied:
Thanks. I think it's fine; if some are missing I'll add a note.

cowtowncoder commented on cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java:
```diff
@@ -46,17 +48,92 @@ public static int collectDefaults()
+            return _tags[0];
+        }
+
+        private int[] _tags;
```
Maybe add `size()` method? (`return _tagCount;`)

here-abarany replied:
I've added the method as requested.

cowtowncoder commented on cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:
```diff
@@ -501,42 +536,6 @@ public final void writeFieldId(long id) throws IOException {
-     * Specialize {@link JsonGenerator#copyCurrentEvent} to handle tags.
-     */
-    @Override
-    public void copyCurrentEvent(JsonParser p) throws IOException {
```
Where are these implemented now? Without overrides tag information will be lost if buffering (although TBH this won't work with general `TokenBuffer` anyway)

here-abarany replied:
This is explained in the message for the first commit in the series:

> Copying tags can be dangerous in two main cases:
> 1. Jackson creates its own tags, such as for BigNum or the self describe
   tag.
> 2. The semantic meaning of an element changes. For example, converting
   between stringrefs and no stringrefs.
>
> In some cases it would result in duplicate tags, while in others it may
cause completely incorrect tags, such as a stringref tag applying to a
string rather than an integer. Before nested tag support was added this
could generate files that Jackson could not parse, though even with nested
tag support it may cause undefined behavior.

cowtowncoder replied:
Ok, not sure about part "Jackson creates its own tags" here (they come from CBOR spec)?

But aside from that, `copyCurrentEvent` is used to copy exact contents of input. I guess it can be problematic for shared String references as unlike other tags that only related to specific token/value without context/state, string-refs do have context (dictionary being collected).

Given that no unit test fails I guess this is fine. And if this becomes a problem can re-implement methods as necessary.

here-abarany replied:
I mean standard CBOR tags that are created based on the generation code within the `CBORGenerator` class.

This behavior is because the event forwarding still goes through the standard generator interface. For example, let's say you are reading a value that gets interpreted as a BigNum. A CBOR tag is created to decorate the binary string value that holds the data to denote that it represents a BigNum. This logic will be executed regardless of whether an existing BigNum tag exists or not, resulting in *two* BigNum tags. (one from the tag copy, one created from `CBORGenerator`)

As more tags are supported by Jackson, the more dangerous that behavior can get. For example, let's say you're converting from a stringref encoded CBOR to one without stringrefs. It would read an integer index corresponding to a string and convert it to a "real" string, but the tag copying logic it would *also* apply the tag denoting that the value is a stringref index. This is invalid as the stringref tag is intended to apply to an integer, while it would apply to a string instead.

cowtowncoder replied:
Ah. Ok, yes.

cowtowncoder commented on cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java:
```diff
@@ -245,6 +322,35 @@ private Feature(boolean defaultState) {
+    /**
+     * Stack of text and binary string references.
+     */
+    protected Stack<StringRefList> _stringRefs = new Stack<>();
```
Could probably create sub-class that has helper methods, if that simplifies calling code (will add a note where this'd help)

cowtowncoder commented on cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java:
```diff
@@ -615,17 +732,25 @@ public JsonToken nextToken() throws IOException
             if (!_streamReadContext.expectMoreValues()) {
-                _tagValue = -1;
+                --_nestedDepth;
+                if (!_stringRefs.empty() && _stringRefs.peek().depth == _nestedDepth) {
```
Using a simple sub-class of `Stack` would allow hiding this logic into single method.

here-abarany replied:
I created a helper class for this. I chose to use containment rather than subclass for this case as the interface is slightly different.

cowtowncoder replied:
@here-abarany Ok yes, that is cleaner approach from OO design perspective. Although being only used internally not a big thing either way.

cowtowncoder commented on cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java:
```diff
@@ -837,6 +978,9 @@ protected String _numberToName(int ch, boolean neg) throws IOException
                 // got some edge case to consider
                 if (i < 0) {
+                    if (isStringref) {
+                        _reportError("String reference index too large.");
```
Please remove trailing `.`

here-abarany replied:
I think I got all the added errors, let me know if I missed any.

cowtowncoder commented:
Ok @here-abarany -- impressive changes! I had some questions etc, although nothing major.

My main concern is actually getting this to merge against 3.0 (`master`). I know it's not practical to split this into smaller chunks, but one thing that might help is if white-space changes (removal of white-space from end-of-lines) could be done as a separate PR first?
I could merge that in, then into `master`; that'd probably remove noise slightly.

I suspect I will need help with `master` merge anyway but maybe little bit less.

here-abarany commented:
@cowtowncoder thanks for the review, I hope to push up an update later today based on your comments. I'll have two separate pushes: one to rebase onto the latest 2.15 changes and one to push up the updates.

With regards to the whitespace changes, my editor is configured to clear trailing whitespace on save. I might be able to check out to the 2.15 head, save those files, and cherry-pick my changes on top of that to reduce some of the noise. If I'm successful I'll include that with my pushes later today.

I can try merging into the master branch locally and see what happens. Hopefully everything merges cleanly, if not I'll note what issues I encountered.

cowtowncoder commented:
> @cowtowncoder thanks for the review, I hope to push up an update later today based on your comments. I'll have two separate pushes: one to rebase onto the latest 2.15 changes and one to push up the updates.
> 
> With regards to the whitespace changes, my editor is configured to clear trailing whitespace on save. I might be able to check out to the 2.15 head, save those files, and cherry-pick my changes on top of that to reduce some of the noise. If I'm successful I'll include that with my pushes later today.
> 
> I can try merging into the master branch locally and see what happens. Hopefully everything merges cleanly, if not I'll note what issues I encountered.

It's unlikely to merge cleanly, partly due to package renaming, partly due to big API changes.
But exactly how tricky it is can vary. :)

As to whitespace, I'll try to figure out how to do that on my end for 2.15; that way you can just rebase on 2.15 head. No need for you to have to do that now I think.

cowtowncoder commented:
@here-abarany Ok I did white-space trimming, merged it to `master` (re-linking some sources that hopefully reduces conflicts -- changing Java package causes lots of pain). So you may want to merge/rebase from 2.15 now.

here-abarany commented:
@cowtowncoder I have pushed up my changes, I believe I addressed all the comment threads from the review. The first push is just a rebase, so you should be able to view just my latest changes by comparing the latest push just before this comment.

cowtowncoder commented on cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:
```diff
@@ -236,6 +253,11 @@ public int getMask() {
 
+    /**
+     * Table of previously referenced text and binary strings.
+     */
```
... Used if StringRef functionality is enabled. // or something like that

`@since 2.15`

here-abarany replied:
I wasn't sure if protected fields should be marked, I'll go through and try to add it for the fields I've added. (as well as the comment clarification noted here)

cowtowncoder commented on cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java:
```diff
@@ -245,6 +331,64 @@ private Feature(boolean defaultState) {
+
+    /**
+     * Shared string that should be used in place of _textBuffer when a string reference is used.
+     */
```
@since 2.15

cowtowncoder commented on cbor/src/test/java/com/fasterxml/jackson/dataformat/cbor/ParserWithJsonOrgSampleTest.java:
```diff
@@ -47,6 +47,16 @@ public void testJsonSampleDoc() throws IOException
 
+    public void testJsonSampleDocStringref() throws IOException
+    {
+        byte[] data = cborDoc(new CBORFactory().enable(CBORGenerator.Feature.STRINGREF),
```
Could you change this to use Builder? This merges cleanly with 3.0 where factories are immutable.

here-abarany replied:
I've switched this and the one below to use a builder.

cowtowncoder commented on cbor/src/test/java/com/fasterxml/jackson/dataformat/cbor/mapper/BiggerDataTest.java:
```diff
@@ -137,4 +140,33 @@ public void testRoundTrip() throws Exception
+		Citm citm0 = MAPPER.readValue(getClass().getResourceAsStream("/data/citm_catalog.json"),
+				Citm.class);
+		ObjectMapper mapper = new CBORMapper(
+		    new CBORFactory().enable(CBORGenerator.Feature.STRINGREF));
```
Same here, builder over factory configuration (I know, many existing tests use this, trying to move over with new code).

cowtowncoder commented:
@here-abarany Ok, sorry, missed couple of smaller things; now done with reviewing I think. Minor test changes requested to make it more likely things merge less uncleanly.

here-abarany commented:
@cowtowncoder I have pushed up a new revision to fix your latest comments.

cowtowncoder commented:
Thank you @here-abarany! I will pick this up again today or tomorrow, get it merged.

cowtowncoder commented:
Ok, phew. Merge to `master` went ok, almost: there is just one new failure and that's for String-Refs; `BiggerDataTest.testRoundTripStringref()`.

I suspect it is since `CBORParser.nextFieldName()` methods (renamed as just `nextName()` in 3.0) do not all handle String-refs.
Or rather, new `nextNameMatch()` method doesn't -- it probably needs to, and I think `jackson-databind` uses that for Bean properties in 3.0.

Could you have a look if you can see how to fix this, @here-abarany ?

here-abarany commented:
@cowtowncoder thanks for getting it merged in, I'll see if I can take a look on Monday for the master issue.

cowtowncoder commented:
> @cowtowncoder thanks for getting it merged in, I'll see if I can take a look on Monday for the master issue.

Thank you in advance! Hopefully it is relatively straight-forward.
